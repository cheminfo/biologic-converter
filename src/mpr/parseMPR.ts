import { BinaryData } from 'cheminfo-types';
import { IOBuffer } from 'iobuffer';

import { parseData, ParseData } from './modules/parseData';
import { parseLogs, ParseLogs } from './modules/parseLogs';
import { parseLoop, ParseLoop } from './modules/parseLoop';
import { parseModuleHeader, ModuleHeader } from './modules/parseModuleHeader';
import { parseSettings, ParseSettings } from './modules/parseSettings';
import { isModule } from './utility/isModule';

export interface MPR {
  name: string /** a string in the first line */;
  data: { header: ModuleHeader; variables: ParseData };
  settings: { header: ModuleHeader; variables: ParseSettings };
  log?: { header: ModuleHeader; variables: ParseLogs };
  loop?: { header: ModuleHeader; variables: ParseLoop };
}

/**
 * Parses binary `.mpr` files
 * arrayBuffer - the data itself.
 * @returns the data as a JSON-like object.
 */
export function parseMPR(arrayBuffer: BinaryData): MPR {
  const buffer = new IOBuffer(arrayBuffer);
  let mpr: Partial<MPR> = {};

  // top level properties
  mpr.name = buffer
    .readUtf8(0x34)
    .replace(/\x1A|\x00/g, '')
    .trim();

  while (isModule(buffer)) {
    const header = parseModuleHeader(buffer); //this is added to the objects below
    const dataStart = buffer.offset;
    const dataLength = header.length;

    if (/settings/i.test(header.longName)) {
      mpr.settings = { header, variables: parseSettings(buffer) };
    } else if (/data/i.test(header.longName)) {
      mpr.data = { header, variables: parseData(buffer, header) };
    } else if (/log/i.test(header.longName)) {
      mpr.log = { header, variables: parseLogs(buffer) };
    } else if (/loop/i.test(header.longName)) {
      mpr.loop = { header, variables: parseLoop(buffer) };
    }
    buffer.offset = dataStart + dataLength; //end of module
  }
  return mpr as MPR;
}
