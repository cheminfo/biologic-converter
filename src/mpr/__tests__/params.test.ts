import { readFileSync } from 'fs';
import { join } from 'path';

import { parseMPR } from '../parseMPR';

const testFiles = '../../__tests__/data/test';


function nanToNaN(objects:Record<string,string|string|number>[]){

objects.forEach((param:Record<string,string|number>, i:number) => {
  for (let [key,val] of Object.entries(param)){
   if(val==="nan") {
    objects[i][key] = NaN
}}})

}

describe('Compare the technique (settings parser)', () => {

  it('ca params', () => {
    const caParams = JSON.parse(
      readFileSync(join(__dirname, `${testFiles}/ca-params.json`), 'utf8')
    );
      // convert to NaN what they put as "nan"
    nanToNaN(caParams);
    const parsed = parseMPR(readFileSync(join(__dirname, `${testFiles}/ca.mpr`)))
    const params = parsed.settings.variables.params
    expect(caParams[0]).toMatchObject(params)
  });

  it('cp params', () => {

    const cpParams = JSON.parse(readFileSync(join(__dirname, `${testFiles}/cp-params.json`), 'utf8'));
    nanToNaN(cpParams)
    const parsed = parseMPR(readFileSync(join(__dirname, `${testFiles}/cp.mpr`)))
    const params = parsed.settings.variables.params 
    expect(cpParams[0]).toMatchObject(params)
  });

  it('cv params', () => {
    const cvParams = JSON.parse(readFileSync(join(__dirname, `${testFiles}/cv-params.json`), 'utf8'));
    nanToNaN(cvParams)
    const parsed = parseMPR(readFileSync(join(__dirname, `${testFiles}/cv.mpr`)))
    const params = parsed.settings.variables.params 
    expect(cvParams[0]).toMatchObject(params)
})
  it('lsv params', () => {
    const lsvParams = JSON.parse(readFileSync(join(__dirname, `${testFiles}/lsv-params.json`), 'utf8'));
    nanToNaN(lsvParams)
    const parsed = parseMPR(readFileSync(join(__dirname, `${testFiles}/lsv.mpr`)))
    const params = parsed.settings.variables.params 
    expect(lsvParams[0]).toMatchObject(params)
  });
  it('wait params', () => {//the python script fails w this file. 
    const json = JSON.parse(readFileSync(join(__dirname, `${testFiles}/WAITmeta.json`), 'utf8'))
    const WAITparams = json.params
    nanToNaN(WAITparams)
    const parsed = parseMPR(readFileSync(join(__dirname, `${testFiles}/wait.mpr`)))
    const params = parsed.settings.variables.params 
    expect(WAITparams[0]).toMatchObject(params)
  });

  it('zirParams', () => {

    const zirParams = JSON.parse(readFileSync(join(__dirname, `${testFiles}/zir-params.json`), 'utf8'));
    nanToNaN(zirParams)
    const parsed = parseMPR(readFileSync(join(__dirname, `${testFiles}/zir.mpr`)))
    const params = parsed.settings.variables.params 
    expect(zirParams[0]).toMatchObject(params)
})
})
