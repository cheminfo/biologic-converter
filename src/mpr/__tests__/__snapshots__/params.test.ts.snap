// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Compare the technique (settings only) cp meta 1`] = `
{
  "acquisitionStart": 0,
  "activeMaterialMass": 0.0010000000474974513,
  "atX": 0,
  "atomicWeight": 0.0010000000474974513,
  "batteryCapacity": 3.615350037958028e-43,
  "batteryCapacityUnit": 0,
  "characteristicMass": 0.0010000000474974513,
  "comments": "",
  "eTransferred": 1,
  "electrodeMaterial": "",
  "electrodeSurfaceArea": 0.0010000000474974513,
  "electrolyte": "",
  "molecularWeight": 0.0010000000474974513,
  "params": {
    "EM": NaN,
    "E_range_max": 10,
    "E_range_min": -10,
    "I_range": "1 A",
    "Is": -100,
    "Is_unit": "mA",
    "Is_vs": 2,
    "bandwidth": 8,
    "dEs": 0,
    "dQM": 3.3329999446868896,
    "dQM_unit": 1,
    "dts": 1,
    "goto_Ns": 0,
    "nc_cycles": 0,
    "record": 1,
    "ts": 120,
  },
  "referenceElectrode": "SCE Saturated Calomel Electrode",
  "technique": "CP",
}
`;
